### DISCUSSION TITLE
How to retrieve back Sequence ID from the successful aligned .vdjca file?

### DISCUSSION THREAD

**Discussion by ShaowenJ** [milaboratory/mixcr#1031] [Q&A] [answered]
Dear MixCR developers,

I have a quick question, do you know how to or in what parameters can I retrieve back the fastq Sequence ID in the alignment step? I know you can get the aligned Sequence by "mixcr exportAlignments", but don't see the sequence ID.
I am using the Nanopore preset.
If there's one, that will be great! Otherwise I could search by the aligned sequence.

Best!

**Comment by mizraelson** [comment, marked as answer]
Hi,
Yes, it is possible to export both read Id and a description from the original FASTQ file(to do that you have to add a parameter in `mixcr align`):

```
mixcr align \
    --species hsa \
    -OsaveOriginalReads= true \
   .
   .
   .
 ```

Then in `mixcr exportAlignments` you can add columns with the reads description from the original fastq file and read Ids:

```
mixcr exportAlignments \ 
    -descrsR1 \
    -descrsR2 \
    -readIds \
   .
   .
   .
```

`-descrR1` and `-descrR2` add fastq original headers for every read, while `-readIds` adds a unique ID number assigned by MiXCR. For read Id there is no need to save original reads but you can't easily backtrack alignments to the original reads by that id.

Also, if you want you can add `-OsaveOriginalReads= true` parameter to the `mixcr analyze` command so you don't have to rerun the alignment:

```
mixcr analyze ont-rna-seq-vdj-full-length \
     --species hsa \
     -Malign.parameters.saveOriginalReads=true \
     sample.fastq.gz \
     sample_result
```

But do keep in mind that this parameter increases the size of `.vdjca` file because it attaches all the original reads.

**Reply by ShaowenJ** [reply]
Ha! Thanks @mizraelson for the answer！

**Reply by ShaowenJ** [reply]
Hi @mizraelson , I have another quick question, If I added `-cloneId` parameter to the `mixcr exportAlignments` command, would the CloneID be the same from the intersection from`clones_TRAD\TRB\IGH\IGL\IGK.tsv`? In other words, they can be matched back from the cloneID, right? And What does the cloneID -1 mean? 

Thanks!

**Reply by mizraelson** [reply]
Hi, `-cloneId` can be exported from `.clna` file which has information on both alignments and clones. And yes, in general you can use cloneId to identify alignments that were assembled into a clone. `-1` means that this particular alignment was not used in a clonotype assembly.

But, for `ont-rna-seq-vdj-full-length` preset the clones are initially assembled by `cdr3` and then extended to the full-length using `mixcr assembleContigs` (that approach seems to have a better yield of full-length clones). Because of that, some clones that you see in `.clna` file (that were assembled by `cdr3`) may be splitted due to mutations in V and J regions which will lead to cloneId reassignment for the whole dataset. That means that clones that you see in the final tsv tables most likely will have new Ids.

**Reply by ShaowenJ** [reply]
Hi @mizraelson, thanks for the information!
I checked the overlap of -cloneId with the union of all cloneIDs from the tsv files, they are totally overlapping except for -1. Is this mean they are not reassigned?

**Reply by ShaowenJ** [reply]
Hi @mizraelson I would share the structure of our library here from the other discussion.
Basically, each sequence should be composed as this diagram: 
**Barcode(14bp)+UMI(7bp)+VV(V: any nucleotide but T)+polyA+cDNA**
The Barcode should be matched from a whitelist file like a single-cell method. I can separate by myself.
Currently, I was using PolyA+cDNA as bulk sequences to run the MiXCR ONT preset and trace back the barcode and UMI information by sequence id (I have a dataframe keep this information). But I see a higher number of clones than the number of UMI for most barcodes. The reason for this I guess is the error rate for longer sequences like cDNA is much higher than the UMI, which generates artifact clones caused by sequence errors. So I should perform some UMI demultiplexing to combine these same or similar UMI.
Do you have any suggestions on how to perform it from MiXCR? That would be great!

Any ideas would be appreciated!

Thanks!

**Reply by mizraelson** [reply]
So if I understand correctly th first 14bp is the cell barcode? can you share the whitelist for that?
VV - two nucleotides (anything but T)
Does the length of the polyA vary or is it always the same?
Then you have the full receptor sequence including 5'UTR?

Also, is it possible to share some test raw file (with no preprocessing) so i can test the preset?

**Reply by ShaowenJ** [reply]
Hi @mizraelson , yes sure. How can I share these with you guys? I prefer by email.
The polyA varies, but I guess I could remove the polyA from alignment and only put the cDNA. 
Theoretically, we hope to recover the full sequence, but in reality, it’s not always the case for our results.

Thanks

**Reply by mizraelson** [reply]
Sure, you can use the following email. 

support@milaboratories.com

Also if the test files are large we have a dedicated server for uploading it.
